docs/design/0000-00-01-architecture-of-storage-engine-pagestorage.md

```diff
@@ -0,0 +1,239 @@
+- GC
+
+
+## Design
+
+Currently, there are three versions of PageStorage (V1, V2, V3). We won't describe the details of the V1/V2 in this article. The V2 design and implementation lead to high write amplification and CPU usage under some scenarios, so we propose the V3 version.
+
+
+### Background
+We found some problems with the V2 version in the actual production environment.
+
+1. **There is a risk of data loss in the meta part**. As long as the `checksum` and `buffer size` fields in a single meta buffer are damaged at the same time, the subsequent buffers will be unavailable. This situation may happen when the disk failure happens or the meta part is changed by an unexpected operation.
+2. **The snapshot of MVCC needs to be optimized**. First, the CPU usage occupied by the snapshot should be reduced. Secondly, the implementation of MVCC structure is too complex to be understood and maintained.
+3. **The GC write amplification in the data part is high and the GC task is too heavy**. Since the data part is composed of append write, `compact gc` is frequently triggered. Besides, the meta part is bound to a data part by "PageFile", meaning that we have to apply `compact gc` to compact the small data in the meta part. The `compact gc` means PageStorage need read all of the valid data from the disk, then rewrite it into a new file. Each round of GC brings additional read and write overhead.
+
+Besides these three problems, we also changed the `lock` implements and the `CRC` implements.
+
+### V3 version
+
+![tiflash-ps-v3-architecture](./images/tiflash-ps-v3-architecture.png)
+
+The V3 version of PageStorage consists of two main components, one is `WALStore` and the other is `BlobStore`.
+
+- WALStore(Write Ahead Log Store): Using the write ahead log file format to manager the meta part.
+- PageDirectory: Provides the function of MVCC. Smaller memory usage and faster speed than V2.
+- BlobStore: Provides an spaces management. Using the address multiplexing to manage the data part.
+
+#### WALStore
+
+WALStore using the fixed block space to manager the meta.
+
+![tiflash-ps-v3-walstore](./images/tiflash-ps-v3-wal-store.png)
+
+WALStore used fixed size of a block to manage meta part.
+
+- A block contains multiple meta records.
+- If a block has some space unused, Also unused space size less than a meta record size. Then using a `padding` to full the unused space.
+
+If crc and meta size are broken at the same time. It is only need to discard the data in a single block, and the neighboring block(the prev block or next block) will be preserved. Because WALStore will read meta from single block, If single block got fault won't effect others. Also more convenient to troubleshoot the cause of meta errors.
+
+WALStore provider two main interfaces:
+
+- **apply**: After `apply`, the meta info will be serialized on disks. This happens after writing datas.
+- **read**: Read serialized meta info from disks. This will happen after `PageStorage` restored.
```

Comment:
```suggestion

WALStore provides two main interfaces:

- **apply**: After `apply`, the meta info will be atomicity serialized on disks. This happens after writing data to BlobStore.
- **read**: Read serialized meta info from disks. This will happen when `PageStorage` is being restored.
```